Dedupe script

alanmcruickshank opened the pull request:
A deduplication script for the V2 artifacts.

In V1, we deduplicate in the query, which is wasteful in computation. In the V2 artifacts, there's no explicit deduplication. This script allows the removal of any duplicate records manually if they ever occur.

The intent of this is particularly useful if users have duplicates from the V1 artifacts upload [which we did].

This script is tested on the tails.com project.

alanmcruickshank committed:
Dedupe V2 artifacts script

alanmcruickshank committed:
qualify with schema and database

alanmcruickshank committed:
qualify drop

alanmcruickshank committed:
qualify insert

NiallRees commented:
Definitely think we need a solution for this, but I'd advocate for a merge into statement instead of an insert in the upload artifacts macro. That way we achieve idempotency and avoid ever needing to worry about this

alanmcruickshank commented:
> I'd advocate for a merge into statement instead of an insert

I'm not sure a merge is helpful here. The docs are very clear that if there are potential duplicates in the source data then the result of a merge is non-deterministic: https://docs.snowflake.com/en/sql-reference/sql/merge.html

Or have a misunderstood what you're suggesting?

alanmcruickshank commented:
One option which would work would just be to put the same `DISTINCT` into the migration script rather than as a seperate operation. In my mind this approach is a bit more general purpose though - in that it could be used as a _cleanup_ operation for both a migration _and_ for an unexpected duplicate upload.

NiallRees commented:
> > I'd advocate for a merge into statement instead of an insert
> 
> I'm not sure a merge is helpful here. The docs are very clear that if there are potential duplicates in the source data then the result of a merge is non-deterministic: https://docs.snowflake.com/en/sql-reference/sql/merge.html
> 
> Or have a misunderstood what you're suggesting?

Can we mitigate that by doing a dedupe in the select clause we use in the merge into?

alanmcruickshank commented:
So I think there are a selection of options which we could use in combination:
1. [and this is what I think you're suggesting] is that we change both the V1 and V2 upload artifacts operations to use a `MERGE`. That way we prevent the creation of _new_ duplicates within V1 and within V2 at the point of upload [potentially causing errors if there are pre-existing duplicates however].
2. We add a `DISTINCT` to the _migration_ script - which would prevent migrating any pre-existing duplicates from V1 to V2 (and these definitely existed for us).
3. We add use a `MERGE` in the migration script to elimitate any artifacts which exist in both V1 _and_ V2 at the point of migration.
4. We provide a deduplication script to strip out any existing duplicates within V1 and/or V2 post upload.
5. If we do enough of these we could probably remove the existing deduplication done at query time by the `deduped` CTE in `stg_dbt__artifacts`.

What combination of 1/2/3/4/5 do you think we should do?

NiallRees commented:
> So I think there are a selection of options which we could use in combination:
> 
> 1. [and this is what I think you're suggesting] is that we change both the V1 and V2 upload artifacts operations to use a `MERGE`. That way we prevent the creation of _new_ duplicates within V1 and within V2 at the point of upload [potentially causing errors if there are pre-existing duplicates however].
> 2. We add a `DISTINCT` to the _migration_ script - which would prevent migrating any pre-existing duplicates from V1 to V2 (and these definitely existed for us).
> 3. We add use a `MERGE` in the migration script to elimitate any artifacts which exist in both V1 _and_ V2 at the point of migration.
> 4. We provide a deduplication script to strip out any existing duplicates within V1 and/or V2 post upload.
> 5. If we do enough of these we could probably remove the existing deduplication done at query time by the `deduped` CTE in `stg_dbt__artifacts`.
> 
> What combination of 1/2/3/4/5 do you think we should do?

I think 1, 4 and 5 make sense and should be the least work?

alanmcruickshank commented:
In that case - let's keep this PR about just 4 - and I'll handle 1 & 5 in a separate one.

alanmcruickshank committed:
Add a V1 dedupe script

alanmcruickshank committed:
Add tests for V1 and V2 dedupe script

alanmcruickshank commented:
@NiallRees in line with our convo above, I've added a V1 dedupe script and added both to the test suite.

I can't test the V1 script live anymore (because I've destroyed my V1 artifacts), but it's a copy of the V2 one which I know works.

Assuming tests pass, this is ready to merge. 👍

NiallRees commented on macros/dedupe_artifacts_v1.sql:
```diff
@@ -0,0 +1,24 @@
+{% set dedupe_results_query %}
+
+    create temporary table {{ src_dbt_artifacts.database }}.{{ src_dbt_artifacts.schema }}.dbt_temp_artifact_table as
+        select distinct * from {{ src_dbt_artifacts }};
```
This is potentially a very expensive operation as it is effectively a group by all columns. If we know that one column represents the uniqueness of a record correctly, we can just do a:
```
select * from {{ src_dbt_artifacts }}
qualify row_number() over (partition by invocation_id, dbt_schema_version order by generated_at asc) = 1
```

which is a lot faster

alanmcruickshank replied:
Good idea. The `generated_at` sort is a bit redundant - because it will be the same for each artifact that's duplicated given it's a property of the artifact not the upload time, but the rest is good.

alanmcruickshank replied:
... in fact. I think the best thing to quantify _uniqueness_ is probably the timestamp itself. Is that a crazy idea?

I'm thinking that in the case of a dbt cloud run, where we want to group together multiple results into the same run, the artifact_run_id isn't unique - there may be several different artifacts with the same id. But the timestamp shows that they're different results. Ditto the V1 artifacts.

NiallRees replied:
I don't see why that should be an issue given that the precision is microseconds. I'm in support!

alanmcruickshank replied:
I've totally been overthinking this - see #103 . It should totally be `command_invocation_id` here.

I've update PR and should be ready to rock.

alanmcruickshank committed:
Remove dedupe in artifacts staging model

alanmcruickshank committed:
Dedupe using a qualify rather than a distinct

alanmcruickshank commented:
btw - I think removing the dedupe step in the artifacts base model makes most sense here [i.e. option 5 above].

alanmcruickshank committed:
Add order by clauses

alanmcruickshank committed:
Switch to dedpue on command_invocation_id #103

alanmcruickshank committed:
Rename run_manifest -> manifest

alanmcruickshank committed:
Update README

NiallRees commented on macros/dedupe_artifacts_v2.sql:
```diff
@@ -0,0 +1,34 @@
+
+        create temporary table {{ artifact_table.database }}.{{ artifact_table.schema }}.dbt_temp_artifact_table as
+            select * from {{ artifact_table }}
+            qualify row_number() over (
```
```suggestion
            qualify rank() over (

```

Thinking about it, should this be a rank? The `src_results_nodes` and `src_manifest_nodes` will and should have multiple entries for a single `command_invocation_id`.

alanmcruickshank replied:
Hmmmmmm - you're totally right on the issue, but I think wrong on the solution. Will take a look a revamp this.

NiallRees replied:
Ah yeah rank wouldn't work. Shouldn't it be as simple as:
```
qualify row_number() over (
    partition by command_invocation_id, node_id
    order by artifact_generated_at asc
) = 1;
```

actually?

alanmcruickshank replied:
read my mind :)

See my most recent commit.

NiallRees replied:
Love the table_key lookup!

alanmcruickshank committed:
Dedupe on the correct granularity

NiallRees commented on macros/dedupe_artifacts_v2.sql:
```diff
@@ -0,0 +1,38 @@
+        
+        truncate {{ artifact_table }};
+
+        insert into {{ artifact_table }} select * from {{ artifact_table.database }}.{{ artifact_table.schema }}.dbt_temp_artifact_table;
```
```suggestion
        create or replace {{ artifact_table }} clone {{ artifact_table.database }}.{{ artifact_table.schema }}.dbt_temp_artifact_table;
```

alanmcruickshank replied:
Good idea 👍

NiallRees commented on macros/dedupe_artifacts_v1.sql:
```diff
@@ -0,0 +1,29 @@
+
+    insert into {{ src_dbt_artifacts }} select * from {{ src_dbt_artifacts.database }}.{{ src_dbt_artifacts.schema }}.dbt_temp_artifact_table;
+
+    drop table {{ src_dbt_artifacts.database }}.{{ src_dbt_artifacts.schema }}.dbt_temp_artifact_table;
```
```suggestion
    create or replace table {{ src_dbt_artifacts }} clone {{ artifact_table.database }}.{{ artifact_table.schema }}.dbt_temp_artifact_table;
```

alanmcruickshank replied:
Good idea 👍  (although in this case we still need the `drop`). I think I've implemented what you _meant_ rather than this exact suggestion.

alanmcruickshank committed:
Clone rather than truncate > insert

alanmcruickshank committed:
fix the copy pasta

alanmcruickshank committed:
cannot clone temporary and non-temporary tables.

alanmcruickshank committed:
Merge branch 'main' into ac/dedupe_operation

NiallRees committed:
Fix clone statement

NiallRees commented on macros/dedupe_artifacts_v1.sql:
```diff
@@ -0,0 +1,28 @@
+{% do run_query(dedupe_results_query) %}
+
+
+{% do log("DONE!", info=True) %}
```
```suggestion
{% do log("Done", info=True) %}

```

NiallRees committed:
Update macros/dedupe_artifacts_v1.sql

NiallRees commented on macros/dedupe_artifacts_v2.sql:
```diff
@@ -0,0 +1,37 @@
+
+{% endfor %}
+
+{% do log("DONE!", info=True) %}
```
```suggestion
{% do log("Done", info=True) %}
```

NiallRees committed:
Update macros/dedupe_artifacts_v2.sql
